### PR TITLE
[FIX] web_editor: fix bugs in the new banner command

### PR DESCRIPTION
**Pull request by dhba-odoo**
**Current behavior before PR:**

- Inserting a banner after already existing content it's impossible to
  go to the next line of the banner.
- Inserting a banner did not set focus inside the banner; instead, focus
  was placed after the banner.
- Performing Ctrl+A and backspace inside a banner would unintentionally remove
  Its first element, which could be a block tag
- When a banner was inserted at the first line and Ctrl+A and backspace
  is performed, nothing happens.

**Desired behavior after PR is merged:**

- It's now possible to navigate to the next line of the banner.
- Now, upon creating a banner, focus is now correctly set inside the
  banner.
- Now performing Ctrl+A and backspace inside the banner will no longer remove its
  first element if it's a block tag.
-  Now, all the content except banner will be selected and removed.

task-3432167

### CODE REVIEW

File: addons/web_editor/static/src/js/editor/odoo-editor/src/commands/deleteBackward.js [hunk 1]
```diff
@@ -100,7 +99,7 @@ HTMLElement.prototype.oDeleteBackward = function (offset, alreadyMoved = false,
             parentEl.oid !== "root" &&
             parentEl.parentElement &&
             !parentEl.parentElement.isContentEditable &&
-            paragraphRelatedElements.includes(this.tagName) &&
+            this.tagName === 'P' &&
```
Comment (shsa-odoo) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/commands/deleteBackward.js:102]: only 'P' not other block tags?

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 2]
```diff
@@ -2283,13 +2288,15 @@ export class Wysiwyg extends Component {
                 const bannerElement = parseHTML(`
                     <div class="o_editor_banner o_not_editable lh-1 d-flex align-items-center alert alert-${alertClass} pb-0 pt-3" role="status" data-oe-protected="true">
                         <i class="fs-4 fa ${iconClass} mb-3" aria-label="${_t(title)}"></i>
-                        <div class="o_editable o_editable_no_shadow w-100 ms-3" data-oe-protected="false">
+                        <div class="o_editable_no_shadow w-100 ms-3" data-oe-protected="false" contenteditable="true">
```
Comment (shsa-odoo) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:2291]: purpose?

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 3]
```diff
@@ -2283,13 +2288,15 @@ export class Wysiwyg extends Component {
...
                             <p><br></p>
                         </div>
                     </div>
                 `).childNodes[0];
+
                 this.odooEditor.execCommand('insert', bannerElement);
-                setSelection(bannerElement.querySelector('.o_editor_banner > div'), 0);
+                setSelection(bannerElement.querySelector('.o_editor_banner > div > p'), 0);
+
```
Comment (shsa-odoo) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:2299]: unnecessary diff.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/commands/deleteBackward.js (hunk 1 above)
Comment (dhba-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/commands/deleteBackward.js:102]: @shsa-odoo  No, only the 'P' tag. otherwise, if the first element inside the banner is a header or any block tag then on backspace it will not convert that tag into a P tag.

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js (hunk 3 above)
Comment (dhba-odoo) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js]: @shsa-odoo Changes Done!

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js [hunk 4]
```diff
@@ -4335,6 +4335,10 @@ export class OdooEditor extends EventTarget {
         ) {
             this.setContenteditableLink(link);
         }
+        const oEditableBanner = closestElement(ev.target, '.o_editable_no_shadow');
+        if (oEditableBanner && !oEditableBanner.isContentEditable) {
+            oEditableBanner.setAttribute('contenteditable', 'true');
+        }
```
Comment (shsa-odoo) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:4341]: IMO, naming it `banner` would be good. Also does our banner will be contenteditable= false all the time? if yes then i thisnk we dont need to check  for this `!oEditableBanner.isContentEditable`. and this part of code will simply look like:
`       const banner = closestElement(ev.target, '.o_editable_no_shadow');
        if (banner) {
            banner.setAttribute('contenteditable', 'true');
        }
`

File: addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js [hunk 5]
```diff
@@ -1843,6 +1843,11 @@ export function splitElement(element, offset) {
         index < offset ? before.appendChild(child) : after.appendChild(child);
         index++;
     }
+    const blockEl = closestBlock(after);
+    if (isShrunkBlock(blockEl)) {
+        const br = document.createElement('br');
+        after.appendChild(br);
+    }
```
Comment (shsa-odoo) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js:1850]: Give some scenario why are we adding this, in form of docString.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js (hunk 4 above)
Comment (dhba-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js]: @shsa-odoo When an already existing record is opened or when we refresh the record, only during that time `contentEditable` will be `false` otherwise, it will always be `true`.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js (hunk 5 above)
Comment (dhba-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js:1852]: @shsa-odoo Done!

**Comment by Zynton** [comment]
@dhba-odoo Before I go into the details of your PR, could you please split your changes into (ideally) one commit per fix?

With everything mixed together:
- it's harder to keep track of what does what
- in the future if we blame a line to see what it does it will be harder to know what it was for
- if we need to revert a change later, we won't be able to just target one commit

Then you can also add tests in each commit for the specific bug it tackles.
Thank you.

**Comment by Zynton** [comment]
Thanks @dhba-odoo. Runbot is red though.

**Comment by dhba-odoo** [comment]
@Zynton I've rebased the PR, and Runbot is now green. Thanks!

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 6]
```diff
@@ -2197,13 +2197,13 @@ export class Wysiwyg extends Component {
                 const bannerElement = parseHTML(`
```
Comment (Zynton) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:2207]: About the commit message. Please rename "Heading 1 placeholder" into "Heading 1 Powerbox hint".

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 7]
```diff
@@ -2197,13 +2197,13 @@ export class Wysiwyg extends Component {
...
                     <div class="o_editor_banner o_not_editable lh-1 d-flex align-items-center alert alert-${alertClass} pb-0 pt-3" role="status" data-oe-protected="true">
                         <i class="fs-4 fa ${iconClass} mb-3" aria-label="${_t(title)}"></i>
-                        <div class="w-100 ms-3" data-oe-protected="false">
+                        <div class="w-100 ms-3" data-oe-protected="false" contenteditable="true">
```
Comment (Zynton) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:2210]: Why is this needed? I don't see anything making the banner not editable in the first place, and we sure don't want this to remain editable when the editor is in readonly mode.
If it's just to fix the Powerbox hint, it's the wrong way to fix it. Is it possible to edit the paragraph before your fix?

File: addons/web_editor/static/tests/banner_tests.js [hunk 8]
```diff
@@ -85,6 +85,20 @@ QUnit.module(
...
+            await nextTick();
+            assert.strictEqual(
+                editable.innerHTML,
+                `<p>Test</p><div class="o_editor_banner o_not_editable lh-1 d-flex align-items-center alert alert-info pb-0 pt-3" role="status" data-oe-protected="true" contenteditable="false">
+                        <i class="fs-4 fa fa-info-circle mb-3" aria-label="Banner Info"></i>
+                        <div class="w-100 ms-3" data-oe-protected="false" contenteditable="true">
+                            <h1 placeholder="Heading 1" class=\"oe-hint\"><br></h1>
+                        </div>
+                    </div><p><br></p>`,
+                "Should Convert Paragraph Tag into Heading Tag Inside Newly Created Banner",
```
Comment (Zynton) [inline addons/web_editor/static/tests/banner_tests.js:156]: Same comment as before about the capitals.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/commands/deleteBackward.js [hunk 9]
```diff
@@ -101,7 +101,7 @@ HTMLElement.prototype.oDeleteBackward = function (offset, alreadyMoved = false,
...
```
Comment (Zynton) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/commands/deleteBackward.js:104]: I wrote a comment on the task about this, I'm not sure about this spec. But if we keep it, this does seem like the right way to go about it.

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 10]
```diff
@@ -1963,7 +1964,16 @@ export class Wysiwyg extends Component {
                 this.$editable.find(containerSelector)[0];
             if (container) {
                 const range = document.createRange();
-                range.selectNodeContents(container);
+                if (!container.hasChildNodes()) {
+                    range.selectNodeContents(container);
+                } else {
+                    let start = container.firstChild;
+                    while((!isVisible(start)) && start.nextSibling) {
```
Comment (Zynton) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js]:
```suggestion
                    while (start.nextSibling && !isVisible(start)) {
```

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 11]
```diff
@@ -1963,7 +1964,16 @@ export class Wysiwyg extends Component {
...
+                        start = start.nextSibling;
+                    }
+                    range.setStart(start, 0);
+                    range.setEnd(...endPos(container.lastChild));
```
Comment (Zynton) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:1975]: What you needed to do to find the start node, how come you don't need to do it to find the end node as well?

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 12]
```diff
@@ -1968,7 +1968,7 @@ export class Wysiwyg extends Component {
                     range.selectNodeContents(container);
                 } else {
                     let start = container.firstChild;
-                    while((!isVisible(start)) && start.nextSibling) {
+                    while((!isVisible(start) || !start.isContentEditable) && start.nextSibling) {
```
Comment (Zynton) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:1971]:
```suggestion
                    while (start.nextSibling && !(isVisible(start) && start.isContentEditable)) {
```

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 13]
```diff
@@ -1963,7 +1964,16 @@ export class Wysiwyg extends Component {
                 this.$editable.find(containerSelector)[0];
             if (container) {
                 const range = document.createRange();
-                range.selectNodeContents(container);
+                if (!container.hasChildNodes()) {
```
Comment (Zynton) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:1967]: You make the fix at the shortcut level, but what if the user right clicked, then clicked "select all"?

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 14]
```diff
@@ -1968,7 +1968,7 @@ export class Wysiwyg extends Component {
                     range.selectNodeContents(container);
                 } else {
                     let start = container.firstChild;
```
Comment (Zynton) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:1970]: Same remark about right-click.

**Review by Zynton — CHANGES_REQUESTED** [review]
Good work. A few remarks to address, detailed in my review.

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js (hunk 7 above)
Comment (dhba-odoo) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:2210]: @Zynton I have implemented changes to ensure that when a banner is created, the cursor will now appear inside the banner. Previously, before the merge of this [PR](https://github.com/odoo/odoo/pull/143362/files) the cursor automatically appeared inside the banner upon its creation. However, this functionality was affected after the removal of 'o_editable.'

Regarding the issue related to the powerbox hint, the problem arose because the selection was set on the 'div' element instead of the 'p' tag.

If we dont want the selection inside the banner when it is created, then their is no need for this changes. Additionally, the line of code 'setSelection(bannerElement.querySelector('.o_editor_banner > div'), 0);' is no longer necessary, as it was responsible for setting the selection inside the banner. Therefore, this line can be removed

File: addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js [hunk 15]
```diff
@@ -1867,6 +1867,12 @@ export function splitElement(element, offset) {
         index < offset ? before.appendChild(child) : after.appendChild(child);
         index++;
     }
+    // e.g.: <p>Test/banner</p> + ENTER <=> <p>Test</p><div class="o_editor_banner>...</div><p><br></p>
```
[the remark this answers is not in the document]
Comment (dhba-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js:1870]: When a user creates a banner like in example, an issue arises where the after element appears empty without a line break `<br>`. Consequently, it becomes challenging to access the next line after the banner if needed.

Similarly, when creating a table using a similar approach, clicking at the end of the table 'type / for commands' powerbox hint will appear below the table. However, the cursor is still positioned at the end of the table, and any typing at that point inserts content outside the paragraph `(<p>)`. like, `<table></table>test<p></p>`
I have check the sanitizer and it's not adding `<br>` in those empty tags.

File: addons/web_editor/static/tests/banner_tests.js [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+                    "</form>",
+                resId: 1,
+            });
+        },
+    },
+    function () {
+
+        QUnit.module("Banner");
+
+        QUnit.test("First element of o_editable is editable", async function (assert) {
```
[the remark this answers is not in the document]
Comment (dhba-odoo) [reply addons/web_editor/static/tests/banner_tests.js:125]: All test cases in this file is related to banner, if its confusing i will change it to something meaningful

File: addons/web_editor/static/tests/banner_tests.js [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+            await nextTick();
+            assert.strictEqual(
+                editable.innerHTML,
+                `<p>Test</p><div class="o_editor_banner o_not_editable lh-1 d-flex align-items-center alert alert-info pb-0 pt-3" role="status" data-oe-protected="true" contenteditable="false">
+                        <i class="fs-4 fa fa-info-circle mb-3" aria-label="Banner Info"></i>
+                        <div class="w-100 ms-3" data-oe-protected="false" contenteditable="true">
+                            <p placeholder=\"Type &quot;/&quot; for commands\" class=\"oe-hint oe-command-temporary-hint\"><br></p>
+                        </div>
+                    </div><p><br></p>`,
+                "Should Insert a Banner with Paragraph Tag",
```
[the remark this answers is not in the document]
Comment (dhba-odoo) [reply addons/web_editor/static/tests/banner_tests.js:142]: I apologize for any confusion. Yes, it is followed by a paragraph.

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js (hunk 13 above)
Comment (dhba-odoo) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:1967]: I overlooked this scenario; i will investigate and address it properly.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (Zynton) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js:1870]: Ok indeed I see that `fillEmpty` (which does what you need here) is indeed called in various places where we need it, and only for one specific case in the sanitizer. Note to self: it would be good to refactor that.
In the meantime, it's ok to leave it here but you should use the util (`fillEmpty`) rather than duplicate the logic.

File: addons/web_editor/static/tests/banner_tests.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (Zynton) [reply addons/web_editor/static/tests/banner_tests.js:125]: I mean it's not easy to tell from this title what it has to do with banner. I expect it to have a link with banner since it's in this file, but the title of this test is puzzling.

File: addons/web_editor/static/tests/banner_tests.js (hunk 17 above)
[the remark this answers is not in the document]
Comment (Zynton) [reply addons/web_editor/static/tests/banner_tests.js:142]: OK then please rephrase :-)

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js (hunk 7 above)
Comment (Zynton) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:2180]: Would it work if you kept your `setSelection` change and removed the `contenteditable="true"` ? I would expect it to if the banner is in the root editable. If not, how come?

File: addons/web_editor/static/tests/banner_tests.js [hunk 18]
```diff
@@ -94,6 +94,34 @@ QUnit.module(
...
+            triggerEvent(editor.editable, "input", { inputType: "deleteContentBackward" });
+            assert.strictEqual(
+                editable.innerHTML,
+                `<div class="o_editor_banner o_not_editable lh-1 d-flex align-items-center alert alert-info pb-0 pt-3" role="status" data-oe-protected="true" contenteditable="false">
+                        <i class="fs-4 fa fa-info-circle mb-3" aria-label="Banner Info"></i>
+                        <div class="w-100 ms-3" data-oe-protected="false" contenteditable="true">
+                            <p><br></p>
+                        </div>
+                    </div><p placeholder=\"Type &quot;/&quot; for commands\" class=\"oe-hint oe-command-temporary-hint\"><br></p>`,
+                "should not remove banner when ctrl+a and backspace are preformed",
```
Comment (Zynton) [inline addons/web_editor/static/tests/banner_tests.js:121]: typo: "performed"

File: addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js [hunk 19]
```diff
@@ -830,15 +830,15 @@ export function getDeepestPosition(node, offset) {
...
+        if ((isVisible(next) || isZWS(next)) && (!isBlock(next) || next.isContentEditable)) {
             // Valid node: update position then try to go deeper.
             if (next !== node) {
                 [node, offset] = [next, direction ? 0 : nodeSize(next)];
             }
             // First switch direction to left if offset is at the end.
             direction = offset < node.childNodes.length;
             next = node.childNodes[direction ? offset : offset - 1];
-        } else if (direction && next.nextSibling && !isBlock(next.nextSibling)) {
+        } else if (direction && next.nextSibling && (!isBlock(next.nextSibling) || !next.isContentEditable)) {
```
Comment (Zynton) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js:841]: Semantics and style so feel free not to listen to me, but I prefer `!(isBlock(next.nextSibling) && next.isContentEditable)`. Most of the time I prefer `!(a && b)` over `!a || !b` because it's more intuitive: what we want is to not have the situation where these two things are true at the same time. It's harder to say it the other way. But that's me.

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js (hunk 7 above)
Comment (dhba-odoo) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:2180]: `<div class="o_editor_banner o_not_editable lh-1 d-flex align-items-center alert alert-${alertClass} pb-0 pt-3" role="status" data-oe-protected="true">` is `contentEditable=False` and because of that `setSelection` will not set selection within the `p` tag.
It seems that `o_not_editable` class is responsible for setting `contentEditable` to false.

Comment (Zynton) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:2180]: Ah thank you I hadn't noticed the `.o_not_editable` on the banner!
The [PR you linked to](https://github.com/odoo/odoo/pull/143362) explicitly adds the `div` to which you applied `contenteditable="true"` to the array returned by `getContentEditableAreas`. Then, `_activateContenteditable` is supposed to apply `contenteditable="true"` to it automatically (and `_stopContenteditable` is supposed to remove it so you wouldn't still have it in readonly). So your code should not be needed. Instead, we need to figure out why `_activateContenteditable` doesn't apply `contenteditable="true"`. When it calls `getContentEditableAreas`, does it indeed receive the banner's div as part of the response? I see the PR has a big comment above `getContentEditableAreas`:
```js
// TODO other places in this file call this.options.getContentEditableAreas
// without the extension here. It does not seem to be a problem (it
// was like that before o_editor_banner elements were considered
// here), but we might want to review that.
```
It seems that this would be the most obvious cause. If `_activateContenteditable` doesn't receive the banner's div as part of the response from `getContentEditableAreas`, then `getContentEditableAreas` is called without the needed extension, and that's what needs fixing.

File: addons/web_editor/static/tests/banner_tests.js [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+            setupViewRegistries();
+            patchWithCleanup(Wysiwyg.prototype, {
+                async init() {
+                    await this._super(...arguments);
+                    wysiwyg = this;
+                },
+            });
+            await makeView({
+                type: "form",
+                serverData,
```
[the remark this answers is not in the document]
Comment (Zynton) [reply addons/web_editor/static/tests/banner_tests.js:52]: By "here" I meant directly in `beforeEach`. I don't see a reason to keep it global.

File: addons/web_editor/static/tests/banner_tests.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (Zynton) [reply addons/web_editor/static/tests/banner_tests.js:123]: So, is the test more like "Insert a banner"?

File: addons/web_editor/static/tests/banner_tests.js [hunk 21]
```diff
@@ -61,8 +61,37 @@ QUnit.module(
     },
     function () {
 
-        QUnit.module("banner insertion tests");
+        QUnit.module("banner selection and backspace related tests");
```
Comment (Zynton) [inline addons/web_editor/static/tests/banner_tests.js:64]: No need to include "tests" in the name.

File: addons/web_editor/static/tests/banner_tests.js [hunk 22]
```diff
@@ -61,8 +61,37 @@ QUnit.module(
...
+            assert.strictEqual(
+                editable.innerHTML,
+                `<p>Test</p><div class="o_editor_banner o_not_editable lh-1 d-flex align-items-center alert alert-info pb-0 pt-3" role="status" data-oe-protected="true" contenteditable="false">
+                        <i class="fs-4 fa fa-info-circle mb-3" aria-label="Banner Info"></i>
+                        <div class="w-100 ms-3" data-oe-protected="false" contenteditable="true">
+                            <p placeholder=\"Type &quot;/&quot; for commands\" class=\"oe-hint oe-command-temporary-hint\"><br></p></div></div><p><br></p>`,
+            );
+        });
 
+        QUnit.module("banner insertion tests");
```
Comment (Zynton) [inline addons/web_editor/static/tests/banner_tests.js:122]: Same.

File: addons/web_editor/static/tests/banner_tests.js (hunk 18 above)
Comment (Zynton) [reply addons/web_editor/static/tests/banner_tests.js:118]: The typo is still there.

**Review by Zynton — CHANGES_REQUESTED** [review]
@dhba-odoo I saw you force pushed a few times two weeks ago but it seems you haven't yet addressed the review comments so I stopped re-reviewing. Are you planning on picking this up again soon?

**Comment by dhba-odoo** [comment]
> @dhba-odoo I saw you force pushed a few times two weeks ago but it seems you haven't yet addressed the review comments so I stopped re-reviewing. Are you planning on picking this up again soon?

Yes, I'm currently working on it. After making changes to the code, I moved it for functional testing to identify any cases that may not be working as intended. Additionally, I've updated the entire test case.

File: addons/web_editor/static/tests/banner_tests.js (hunk 21 above)
Comment (dhba-odoo) [reply addons/web_editor/static/tests/banner_tests.js]: done!

File: addons/web_editor/static/tests/banner_tests.js (hunk 22 above)
Comment (dhba-odoo) [reply addons/web_editor/static/tests/banner_tests.js]: done!

File: addons/web_editor/static/tests/banner_tests.js (hunk 20 above)
[the remark this answers is not in the document]
Comment (dhba-odoo) [reply addons/web_editor/static/tests/banner_tests.js:52]: I have implemented the suggested changes!

File: addons/web_editor/static/tests/banner_tests.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (dhba-odoo) [reply addons/web_editor/static/tests/banner_tests.js]: Yes, it's to check whether the new banner is properly inserted or not.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (dhba-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js:1871]: I've applied the suggested changes.

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js (hunk 7 above)
Comment (dhba-odoo) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js]: In a typical scenario, `_activateContenteditable` is called on `_onMouseDown` or `historyRevertCurrentStep`. However, when creating a new banner, neither of these methods is invoked. Consequently, `_activateContenteditable` is not called, and the banner remains non-editable upon creation. To address this, I have call `_activateContenteditable` after the banner is inserted to ensure it make the banner editable.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js (hunk 19 above)
Comment (dhba-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js]: I have removed these specific changes, so there is no longer a need to modify it.

File: addons/web_editor/static/tests/banner_tests.js (hunk 18 above)
Comment (dhba-odoo) [reply addons/web_editor/static/tests/banner_tests.js]: done!

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 23]
```diff
@@ -2203,7 +2203,8 @@ export class Wysiwyg extends Component {
                     </div>
                 `).childNodes[0];
                 this.odooEditor.execCommand('insert', bannerElement);
-                setSelection(bannerElement.querySelector('.o_editor_banner > div'), 0);
+                this.odooEditor._activateContenteditable();
+                setSelection(bannerElement.querySelector('.o_editor_banner > div > p'), 0);
```
Comment (shsa-odoo) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:2207]: sorry for the last minute input, i observed something which should be addressed as well. @dhba-odoo accessing the private method is not recommended, i have a suggestion which can be a workaround 
 1) by using 't-ref' to banner element and setting focus to it's <p>. 
2) or we can make '_activateContenteditable' a public method keeping it in utils. such that `wysiwyg` and `odooeditor` both can use it according to its convenience.
cc @Zynton what is your input on this ?

Comment (Zynton) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:2207]: It's indeed not ideal and I'm a little confused as to why it's being accessed anyway if the goal is just to set the focus. Wouldn't `bannerElement.focus()` suffice? If not, how come?

Comment (dhba-odoo) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:2207]: Here, `bannerElement.focus()` will not work because the `contentEditable` attribute for the`bannerElement` is `false`.

Comment (Zynton) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:2207]: Ah I see that `contenteditable` is only set to true on the `div` that you want to edit (`.o_editor_banner > div`) because it's explicitely added as a content editable area:
```js
getContentEditableAreas: (...args) => {
    const areaEls = _getContentEditableAreas?.(...args) || [];
    const bannerEls = this.$editable[0].querySelectorAll('.o_editor_banner > div');
    return [...areaEls, ...bannerEls];
},
```
So indeed that only gets applied when `_activateContenteditable` is called, which means we need to make sure that gets called first... You could trigger a `mousedown` event for that but that's unclear and you'd have to deal with asynchronous stuff, so I'd argue in favor of making `_activateContenteditable` public. Although we can't change its name in stable so my suggestion is to add a method:
```js
activateContenteditable {
    this._activateContenteditable();
}
```
with a nice comment to justify it.

@sebgeelen is that ok for you?

Comment (sebgeelen) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js]: sure

Comment (dhba-odoo) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js]: Done!

File: addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js [hunk 24]
```diff
@@ -1868,6 +1868,11 @@ export function splitElement(element, offset) {
         index < offset ? before.appendChild(child) : after.appendChild(child);
         index++;
     }
+    // e.g.: <p>Test/banner</p> + ENTER <=> <p>Test</p><div class="o_editor_banner>...</div><p><br></p>
+    const blockEl = closestBlock(after);
+    if (isShrunkBlock(blockEl)) {
```
Comment (Zynton) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js:1873]: `fillEmpty` tests this too. Do you need to do it here as well so as not to insert zws? Otherwise you can remove this `if`.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js [hunk 25]
```diff
@@ -2011,6 +2011,15 @@ export class OdooEditor extends EventTarget {
...
 
+    /**
+     * `activateContenteditable` serves as an interface for external use,
+     * allowing users to conveniently trigger `_activateContenteditable`
+     * from outside the odooEditor.
+     */
+    activateContenteditable() {
+        this._activateContenteditable();
+    }
+
```
Comment (Zynton) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:2022]: Note: This should be adapted in master so we just rename the private function instead of duplicating.

File: addons/web_editor/static/tests/banner_tests.js [hunk 26]
```diff
@@ -0,0 +1,85 @@
...
+                    '<field name="body" widget="html" style="height: 100px"/>' +
+                    "</form>",
+                resId: 1,
+            });
+        },
+    },
+    function () {
+
+        QUnit.module("insert a new banner");
+        QUnit.test("should insert a banner with paragraph tag", async function (assert) {
```
Comment (Zynton) [inline addons/web_editor/static/tests/banner_tests.js:65]: Didn't you say you meant "followed by a paragraph" rather than "with paragraph"? The distinction matters because "with paragraph" could mean that what matters in your test is that the banner _contains_ a paragraph (which it does), but from what I understand you want to make sure the banner is inserted and _also_ that a paragraph is inserted _after it_.

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js (hunk 6 above)
Comment (Zynton) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js]: Is this about a commit that disappeared?

**Review by Zynton — CHANGES_REQUESTED** [review]
Not an easy series of fixes but you're almost there!

File: addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js (hunk 24 above)
Comment (dhba-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js]: when I initially added the `fillEmpty` function, some test cases failed during local testing due to `zws` but now its working correctly. Therefore, it can be removed.

File: addons/web_editor/static/tests/banner_tests.js (hunk 26 above)
Comment (dhba-odoo) [reply addons/web_editor/static/tests/banner_tests.js]: I thought I changed it, but I probably accidentally reverted it while making other changes. Sorry for the inconvenience.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js (hunk 24 above)
Comment (Zynton) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js]: I checked and there is no reason we should ever enter the `if` that adds `zws` in this case since we're passing it a block and if it's shrunk it will add a `<br>`, making it visible. If it's not shrunk, there's no reason it would not be visible. So this change is good, thanks.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js [hunk 27]
```diff
@@ -1868,6 +1868,9 @@ export function splitElement(element, offset) {
         index < offset ? before.appendChild(child) : after.appendChild(child);
         index++;
     }
+    // e.g.: <p>Test/banner</p> + ENTER <=> <p>Test</p><div class="o_editor_banner>...</div><p><br></p>
+    const blockEl = closestBlock(after);
+    blockEl && fillEmpty(blockEl);
```
[the remark this answers is not in the document]
Comment (dhba-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js]: done!

**Comment by Zynton** [comment]
@robodoo rebase-ff r+

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js (hunk 25 above)
Comment (Zynton) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:2022]: Done here: https://github.com/odoo/odoo/pull/152825
